ext/stackprof/stackprof.c

```diff
@@ -536,31 +571,39 @@ stackprof_record_gc_samples()
 static void
 stackprof_gc_job_handler(void *data)
 {
-    static int in_signal_handler = 0;
-    if (in_signal_handler) return;
+    if (_stackprof.in_signal_handler) return;
     if (!_stackprof.running) return;
 
-    in_signal_handler++;
+    _stackprof.in_signal_handler++;
     stackprof_record_gc_samples();
-    in_signal_handler--;
+    _stackprof.in_signal_handler--;
 }
 
 static void
 stackprof_job_handler(void *data)
 {
-    static int in_signal_handler = 0;
-    if (in_signal_handler) return;
+    if (_stackprof.in_signal_handler) return;
     if (!_stackprof.running) return;
 
-    in_signal_handler++;
+    _stackprof.in_signal_handler++;
     stackprof_record_sample();
-    in_signal_handler--;
+    _stackprof.in_signal_handler--;
 }
 
 static void
 stackprof_signal_handler(int sig, siginfo_t *sinfo, void *ucontext)
 {
     _stackprof.overall_signals++;
+
+    // Protect against individual samples taking longer to capture than
+    // the interval between samples, which would cause the job queue
+    // to pile up faster than it's flushed, peg the CPU, and hang the program.
+    if (_stackprof.in_signal_handler) {
+        if (_stackprof.debug)
+            fprintf(stderr, "skip stackprof_signal_handler, already in handler!\n");
+        return;
```

Comment:
Another action we could consider doing in this case, is injecting fake frames, similar to GC, named `<skip>` or similar, so that the final report makes it clear that the sampling rate was too fast, rather than being silently lossy.